README.md
# Infernyx

[![Build Status](https://travis-ci.org/mozilla/infernyx.svg?branch=master)](https://travis-ci.org/mozilla/infernyx)

This project contains data processing rules for the Disco and Inferno mapreduce rules for the Mozilla Tiles project.

Infernix Input Payloads
-----------------------

The payload it expects is:

    {
      "timestamp": 1407336655489, # unix timestamp, injected by onyx
      "date": "2014-05-27", # iso formatted date string for easy splitting by date, injected by onyx
      "ip": "103.242.154.10", # request originator's IPv4 address, injected by onyx
      "ua": "Mozilla/5.0 (Windows NT 5.1; rv:33.0) Gecko/20100101 Firefox/33.0", # request originator's UA, injected by onyx
      "locale": "en-US", # locale str, sent by client
      "click": 2, # this denotes a click action, and the index of the tile click in the "tiles" array. other possible actions: "block", "pin", "unpin"
      "tiles": [
        {
          "url": "http://stackoverflow.com/questions/5998245/get-current-time-in-milliseconds-in-python", # url if a history tile, absent otherwise
          "id": 8, # tile id if it is an enhanced tile or directory tile. absent otherwise
          "pin": true # if the tile is pinned, absent otherwise
        },
        ...
      ]
    }

Payloads from Firefox
---------------------

Valid request examples from firefox, one per line:

Impression ping, because there are no actions

    {"locale":"en-US","tiles":[{"id":1},{"id":2},{"id":3},{"id":4},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}]}

Click ping, the action being "pin"

    {"locale":"en-US","tiles":[{"id":1,"pin":1},{"id":2},{"id":3},{"id":4},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}],"pin":0}

Click ping, the action being "unpin"

    {"locale":"en-US","tiles":[{"id":1},{"id":2},{"id":3},{"id":4},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}],"unpin":0}

Click ping, the action being "block"

    {"locale":"en-US","tiles":[{"id":1},{"id":2},{"id":3},{"id":4},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}],"block":0}

Impression ping, after the "block" action

    {"locale":"en-US","tiles":[{"id":2},{"id":3},{"id":4},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}]}

Impression ping, with an enhanced history tile

    {"locale":"en-US","tiles":[{"id":5,"url":"https://www.mozilla.org/en-US/about/"},{"id":2},{"id":3},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}]}

Impression ping, with an enhanced and a regular history tile

    {"locale":"en-US","tiles":[{"id":5,"url":"https://www.mozilla.org/en-US/about/"},{"url":"https://twitter.com/Firefox"},{"id":2},{"id":3},{"id":6},{"id":7},{"id":8},{"id":9},{"id":10}]}

Note the empty tile at position 7

    {"locale":"en-US","tiles":[{"id":5,"url":"https://www.mozilla.org/en-US/about/"},{"id":2},{"id":3},{"id":6},{"id":8},{"id":9},{"id":10},{"pin":1,"pos":8,"url":"https://twitter.com/Firefox"}]}

The web server listening to requests at the api endpoints will then inject a couple of parameters in the log. These include:
 * timestamp: a unix timestamp at time of receipt, in UTC
 * date: an iso formatted date string for easy splitting by date, in UTC
 * ip: request originator's IPv4 address
 * ua: request originator's UA

An important part of the payload sent by firefox is the `tiles` array. It contains a sequence of JSON objects with data about what was clicked.
In the case of a non-impression payload, there will be an action included, which manifests itself as a top-level key to the payload. The value to this key is an integer. This represents the array index in the `tiles` array that this action is about.

e.g.

    {"locale":"en-US","tiles":[{"id":2}],"click":0}

Means that a user clicked on tile at position 0 in `tiles`. That tile happens to be a directory tile with `id` 2.

__Note__: There can only be one action key

This key can be one of:
 * click
 * pin
 * unpin
 * block

## Tile Object

The payload from Firefox includes the `tiles` key which has as value an array of tile objects. This section describes the schema for tile objects found in this array.

These are the parameters a tile object can have:

 * id
 * url
 * pin
 * pos

### id
Value: Integer  
A tile id.

This is sent if the tile is a directory tile or an enhanced history tile. If it is strictly a history tile, this key won't be included in the tile object.

### url
Value: String  
The url of the tile if it is a history tile.

If it isn't a history tile (i.e. it is a directory tile), this key is not included in the tile object.

### pin
Value: Integer  
Status about this tile

If this tile is pinned in firefox, the value of this key will be `1`. otherwise it won't be included in the tile object.

### pos
Value: Integer  
The position of the tile in the new tab page

this parameter is only included if the index in the array is not a meaningful representation of the tiles.
If all tiles are contiguous in the Firefox new tab page (i.e. there are no gaps), this won't be included in the tile object.

e.g. if a user only has 2 tiles, but happens to have dragged a tile to position 9 and has pinned it, there would be a gap between the tile at index 0 and the last tile, which would be at index 1, but ends up being at position 9.

imagine the user having those tiles:

    tiles: [{id: 1}, {id: 2}]

the user drags tile id 2, and moves it to position 9

    tiles: [{id: 1}, {id: 2, pos: 9}]


CHANGELOG.md
0.2.22
======
* Increase the value range for `tile_id` for Pocket

0.2.21
======
* process loaded ping for assa_impression_stats_daily

0.2.20
======
* drop the filter on `source` for assa_impression_stats_daily
* store `receive_at` for assa_impression_stats_daily

0.2.19
======
* stop the duplication of activity stream session pings

0.2.18
======
* filter out PERSONALIZATION_V1_ITEM_RELEVANCE_SCORE_DURATION for activity stream performance pings

0.2.17
======
* process profile_creation_date for assa_events_daily

0.2.16
======
* treat client_id as impression_id if it's present in activity stream router pings

0.2.15
======
* fix parse_locale

0.2.14
======
* process watchdog-proxy event ping

0.2.13
======
* process topsites_search_shortcuts for assa_sessions_daily

0.2.12
======
* add maxerror to redshift LOAD

0.2.11
======
* hotfix to handle the invalid locales

0.2.10
======
* hotfix to handle the "null" addon_version in assa_masga

0.2.9
=====
* hotfix to handle the invalid value for load_trigger_type

0.2.8
=====
* process profile_creation_date for ping_centre_main

0.2.7
=====
* process activity stream router ping

0.2.6
=====
* disable firefox onboarding session2 job

0.2.5
=====
* handle database connection errors

0.2.4
=====
* fix logging in database.py

0.2.3
=====
* optimize result_processor in database.py

0.2.2
=====
* process value for assa_events_daily

0.2.1
=====
* add support for multi-table inserts
* add a new destination table for assa_sessions_daily

0.2.0
=====
* fix the manifest naming issue in result processor 

0.1.99
======
* fix an issue introduced in 0.1.98

0.1.98
======
* use unix gzip to replace the python gzip
* use multifile COPY for AWS Redshift

0.1.97
======
* fix uuid validation so that it recognizes "n/a"

0.1.96
======
* clean up tmp files upon exception in result_processor
* process custom_screenshot, topsites_pinned, profile_creation_date, region for activity stream

0.1.95
======
* add uuid validations

0.1.94
======
* process firefox onboarding v2

0.1.93
======
* check invalid timestamps for firefox onboarding

0.1.92
======
* gzip the database result processor

0.1.91
======
* use ujson in the result processor for better performance

0.1.90
======
* process various new columns for firefox onboarding

0.1.89
======
* process impression_id for AS impression stats
* remove rules for activity stream test pilot and test pilot

0.1.88
======
* process no_image for AS session

0.1.87
======
* process is_preloaded, is_prerendered, topsites_data_late_by_ms, highlights_data_late_by_ms, and topsites_icon_stats from AS session
* add hour and minute for AS impression_stats

0.1.86
======
* process ping_centre_main

0.1.85
======
* process shield_id for activity stream tables
* do not overwrite "version" if provided

0.1.84
======
* exclude all the non-topstories impressions for activity stream system addon

0.1.83
======
* remove tour_source from firefox onboarding events

0.1.82
======
* processs session_begin, session_end, and tour_source for firefox onboarding

0.1.81
======
* process release_channel for activity stream tables
* process firefox onboarding pings

0.1.80
======
* process impression pings for activity stream system addon

0.1.79
======
* process user_prefs for activity stream system addon pings

0.1.78
======
* handle topsites_first_painted_ts for assa_session pings

0.1.77
======
* handle double fields to avoid precision loss for assa_session pings

0.1.76
======
* handle the nested "perf" object for assa_session pings

0.1.75
======
* add a new rule to process pings from activity stream system addon

0.1.74
======
* add support for batch mode payloads and enable it with Activity Stream and Ping-centre

0.1.73
======
* save experiment_id to activity_stream_impression_daily

0.1.72
======
* process impression logs for activity stream

0.1.71
======
* process "topsites_lowresicon" and "topsites_pinned" for activity stream

0.1.70
======
* process "topsites" stats and "user_prefs" for activity stream

0.1.69
======
* treat the "raw" field as optional in Test Pilot pings

0.1.68
======
* fix the missing "event_id" in the masga ping

0.1.67
======
* fix the "null" values in the optional fields

0.1.66
======
* fix the "raw" filed for Test Pilot pings

0.1.65
======
* rename "raw" to "raw_ping" for Test Pilot pings
* add "release_channel" to Activity Stream mobile

0.1.64
======
* process testpilots pings from Ping-Centre
* process Activity Stream Mobile pings from Ping-Centre

0.1.63
======
* fix float values in the performance ping of activity stream

0.1.62
======
* fix import logic for required_modules

0.1.61
======
* bugfix for 2^31 maximums for signed ints in redshift


0.1.59
======
* populate the default value for the `event_id` field in `masga`

0.1.58
======
* process highlights_size for activity stream session pings
* process undesired states pings for activity stream

0.1.57
======
* process tp_version for activity stream shield study

0.1.56
======
* process shield study pings for activity stream

0.1.55
======
* add integer overflow cleaning to activity stream 

0.1.54
======
* process and persist the metadata_source for activity stream

0.1.53
======
* process and persist the share_provider and highlight_type for activity stream

0.1.52
======
* process and persist the recommendation ping for activity stream

0.1.51
======
* add a keyset for application_stats to process activity_stream_performance pings

0.1.50
======
* include the expire_ddfs_da.py into the shipping script

0.1.49
======
* add a script to expire data in ddfs

0.1.48
======
* hotfix: wrap module level funcitons into filters

0.1.47
======
* add a new keyset "activity_stream_event" for rule application_stats
* rename the keyset "activity_stream_stats" to "activity_stream_session"

0.1.46
======
* fix the missing country_code in activity_stream_stats
* add load_latency to activity_stream_stats
* change tab_id type from str to int

infernyx/reload.py
from inferno.lib.rule import chunk_json_stream
from inferno.lib.rule import InfernoRule
from inferno.lib.rule import Keyset
from infernyx.database import insert_postgres, insert_redshift
from infernyx.rules import impression_stats_init, parse_date, parse_locale, parse_ip, parse_ua, \
    combiner, RULES, clean_data, parse_tiles
from functools import partial
from config_infernyx import *

AUTO_RUN = False


def filter_enhanced(parts, params):
    if parts.get('enhanced'):
        yield parts


def count(parts, params):
    parts['count'] = 1
    yield parts

imp_rule = [r for r in RULES if r.name == 'impression_stats'][0]

# {"ver":"2","locale":"he-IL","ip":"37.46.38.244","date":"2014-12-03","timestamp":1417651151217,
# "action":"fetch_locale_unavailable",
# "ua":"Mozilla\/5.0 (X11; Ubuntu; Linux i686; rv:34.0) Gecko\/20100101 Firefox\/34.0"}

RULES = [
    # This rule will give enhanced
    InfernoRule(
        name='enhanced_stats',
        source_tags=['incoming:impression'],
        day_range=1,
        map_input_stream=chunk_json_stream,
        map_init_function=impression_stats_init,
        parts_preprocess=[clean_data, parse_date, parse_locale, parse_ip, parse_ua, parse_tiles, filter_enhanced],
        geoip_file=GEOIP,
        partitions=32,
        sort_buffer_size='25%',
        locale_whitelist={'ach', 'af', 'an', 'ar', 'as', 'ast', 'az', 'be', 'bg', 'bn-bd', 'bn-in', 'br', 'bs',
                          'ca', 'cs', 'csb', 'cy', 'da', 'de', 'el', 'en-gb', 'en-us', 'en-za', 'eo', 'es-ar',
                          'es-cl', 'es-es', 'es-mx', 'et', 'eu', 'fa', 'ff', 'fi', 'fr', 'fy-nl', 'ga-ie', 'gd',
                          'gl', 'gu-in', 'he', 'hi-in', 'hr', 'hu', 'hsb', 'hy-am', 'id', 'is', 'it', 'ja',
                          'ja-jp-mac', 'ka', 'kk', 'km', 'kn', 'ko', 'ku', 'lij', 'lt', 'lv', 'mai', 'mk', 'ml',
                          'mr', 'ms', 'my', 'nb-no', 'nl', 'nn-no', 'oc', 'or', 'pa-in', 'pl', 'pt-br', 'pt-pt',
                          'rm', 'ro', 'ru', 'si', 'sk', 'sl', 'son', 'sq', 'sr', 'sv-se', 'sw', 'ta', 'te', 'th',
                          'tr', 'uk', 'ur', 'vi', 'xh', 'zh-cn', 'zh-tw', 'zu'},
        combiner_function=combiner,
        keysets={
            'impression_stats': Keyset(
                key_parts=['date', 'locale', 'tile_id', 'country_code'],
                value_parts=['impressions', 'clicks', 'pinned', 'blocked', 'sponsored', 'sponsored_link'],
            ),
        },
    ),
    InfernoRule(
        name='application_stats',
        source_tags=['incoming:app'],
        day_range=1,
        map_input_stream=chunk_json_stream,
        map_init_function=impression_stats_init,
        parts_preprocess=[parse_date, parse_ip, parse_ua, count],
        geoip_file=GEOIP,
        partitions=32,
        sort_buffer_size='25%',
        combiner_function=combiner,
        key_parts=['date', 'locale', 'ver', 'country_code', 'action'],
        value_parts=['count'],
    ),
]



